### ISSUE TITLE
Unusual packet during discovery time [8994]

### ISSUE THREAD

**Issue by DungChau1996** [eProsima/Fast-DDS#1093]
Hi all, 

I have been investigating discovery time and behaviour of fastrtps using wireshark. I have observed some behaviour that I did not expect. This can be because I do not thoroughly understand the discovery mechanism. Thus, I hope it is alright to discuss on here. 

About the scenario, I have one publisher and one subscriber communicate with each other. I ran both of them on the same VM. Let's consider only EDP of Data Writer. I receive the first Heartbeat, which I also expected, with:

- `firstAvailableSeqNumber`: 1
- `lastSeqNumber`: 1
- `count`: 1

But then I received 5 more Heartbeat scattering all over the discovery time.
<img width="915" alt="eprosima_discover" src="https://user-images.githubusercontent.com/49442242/77426483-dc9d1d80-6dd4-11ea-8390-9614cfcfe0a2.PNG">

My questions are:

1. Is that a normal behaviour?
2. There is also Heartbeat with `firstAvailableSeqNumber` is1 but `lastSeqNumber` is 0. Is that preemptive Heartbeat? 

Thanks in advance! I apologize if this question is not supposed to be here.

**Comment by DungChau1996** [comment]
I want to add up with unusual packet problem. According to the picture below, the ACKNACK (packet no.121) for the Data(w) (packet no.87) is sent way too slow that make the Discovery duration more than 1s.  
![image](https://user-images.githubusercontent.com/49442242/78259111-83339d80-74fc-11ea-966a-09cd0873df25.png)

This happens very often. Is it a bug in RTPS discovery implementation?

**Comment by EduPonz** [comment]
Hi @DungChau1996 ,

Sorry for the late reply. I'll try to address all three questions.

1. Regarding the Heartbeat count, it's difficult to see what's really happening without the full pcapng file, but it seems reasonable. You are monitoring BUILTIN_PUBLICATION_WRITER entities. There are 2 of such entities, one for the publisher and another one for the subscriber. As you say, there is an intial HEARTBEAT (to trigger ACKNACK messages in the other side), and then there is one HEARTBEAT for each ACKNACK received, which are one preemptive, and at least one more acknowledging the DATA(w) message. That adds up to a total of 3 HEARTBEATS from the publisher and 2 from the Subscriber (since it has no data writers).
1. You are correct, the combination of `firstAvailableSeqNumber = 1` and `lastSeqNumber = 0` is a trick to send and inital HEARTBEAT so that the receiver can send an ACKNACK stating that it's waiting for sequence number 1. This is a redundancy over the preemptive ACKNACK mechanism.
1. Regarding the slow ACKNACK reply for the DATA(w), we saw some time ago a similar behaviour that was corrected on master and 2.0.x branches, I suggest you give those a try. In any case, delaying the ACKNACK does not need to delay EDP discovery, since that happens when the DATA(w/r) is received.

I hope that sheds some light!
